torch/csrc/utils/variadic.h

```diff
@@ -108,4 +108,50 @@ struct MakeIndices : MakeIndices<N-1, N-1, Is...> {};
 // 0 through N-1.
 template<size_t... Is>
 struct MakeIndices<0, Is...> { using indices = Indices<Is...>; };
+
+//===----------------------------------------------------------------------===//
+//                                 Utilities
+//===----------------------------------------------------------------------===//
+
+template <bool value, typename T = void>
+using enable_if_t = typename std::enable_if<value, T>::type;
+
+template <bool value, typename T = void>
+using disable_if_t = enable_if_t<!value, T>;
+
+template <typename T>
+using decay_t = typename std::decay<T>::type;
+
+namespace detail {
+template <bool...>
+struct pack;
+} // namespace detail
+
+template <bool... values>
+struct all_of : std::is_same<
+                    detail::pack<values..., true>,
+                    detail::pack<true, values...>> {};
+
+template <bool...>
+struct any_of_t;
+
+template <>
+struct any_of_t<> : std::false_type {};
+
+template <bool head, bool... tail>
+struct any_of_t<head, tail...> {
+  static constexpr bool value = head || any_of_t<tail...>::value;
+};
+
+template <bool... values>
+struct none_of_t {
+  static constexpr bool value = !any_of_t<values...>::value;
+};
+
+template <bool... values>
+using enable_if_all_of_t = enable_if_t<all_of<values...>::value>;
+
+template <typename T, typename... Ts>
+using disable_for_t =
+    enable_if_all_of_t<(!std::is_same<T, decay_t<Ts>>::value)...>;
```

Comment:
Let's add some (static_assert based) test cases for these type traits. To save compile time, probably rather in a test.cpp file and not in the header.